predicators/spot_utils/perception/object_detection.py

```diff
@@ -0,0 +1,513 @@
+for instructions on setting up DETIC and SAM on a server.
+
+Object detection returns SE3Poses in the world frame but only x, y, z positions
+are currently detected. Rotations should be ignored.
+"""
+
+import io
+import logging
+from pathlib import Path
+from typing import Any, Collection, Dict, List, Optional, Set, Tuple
+
+import apriltag
+import cv2
+import dill as pkl
+import numpy as np
+import PIL.Image
+import requests
+from bosdyn.client import math_helpers
+from matplotlib import pyplot as plt
+from scipy import ndimage
+
+from predicators.settings import CFG
+from predicators.spot_utils.perception.perception_structs import \
+    AprilTagObjectDetectionID, LanguageObjectDetectionID, ObjectDetectionID, \
+    RGBDImageWithContext, SegmentedBoundingBox
+from predicators.utils import rotate_point_in_image
+
+
+def detect_objects(
+    object_ids: Collection[ObjectDetectionID],
+    rgbds: Dict[str, RGBDImageWithContext],  # camera name to RGBD
+) -> Tuple[Dict[ObjectDetectionID, math_helpers.SE3Pose], Dict[str, Any]]:
+    """Detect object poses (in the world frame!) from RGBD.
+
+    Each object ID is assumed to exist at most once in each image, but can
+    exist in multiple images.
+
+    The second return value is a collection of artifacts that can be useful
+    for debugging / analysis.
+    """
+
+    # Collect and dispatch.
+    april_tag_object_ids: Set[AprilTagObjectDetectionID] = set()
+    language_object_ids: Set[LanguageObjectDetectionID] = set()
+    for object_id in object_ids:
+        if isinstance(object_id, AprilTagObjectDetectionID):
+            april_tag_object_ids.add(object_id)
+        else:
+            assert isinstance(object_id, LanguageObjectDetectionID)
+            language_object_ids.add(object_id)
+    detections: Dict[ObjectDetectionID, math_helpers.SE3Pose] = {}
+    artifacts: Dict[str, Any] = {"april": {}, "language": {}}
+
+    # There is no batching over images for april tag detection.
+    for rgbd in rgbds.values():
+        img_detections, img_artifacts = detect_objects_from_april_tags(
+            april_tag_object_ids, rgbd)
+        # Possibly overrides previous detections.
+        detections.update(img_detections)
+        artifacts["april"].update(img_artifacts)
+
+    # There IS batching over images here for efficiency.
+    language_detections, language_artifacts = detect_objects_from_language(
+        language_object_ids, rgbds)
+    detections.update(language_detections)
+    artifacts["language"] = language_artifacts
+
+    return detections, artifacts
+
+
+def detect_objects_from_april_tags(
+    object_ids: Collection[AprilTagObjectDetectionID],
+    rgbd: RGBDImageWithContext,
+    fiducial_size: float = CFG.spot_fiducial_size,
+) -> Tuple[Dict[ObjectDetectionID, math_helpers.SE3Pose], Dict[str, Any]]:
+    """Detect an object pose from an april tag.
+
+    The rotation is currently not detected (set to default).
```

Comment:
also re-mention that the second return value is some 'artifacts' useful for debugging. In this case, it will be the raw apriltag detection results (I think?).